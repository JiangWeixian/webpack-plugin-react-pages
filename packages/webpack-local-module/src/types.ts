export interface ResolvedOptions {
  root: string
  dirname: string
  moduleIdToPath: Record<string, string>
  moduleIdToContent: Record<string, string>
}
