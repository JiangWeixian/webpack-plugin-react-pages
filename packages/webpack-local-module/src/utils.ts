import { withoutLeadingSlash } from 'ufo'

export function resolveId(moduleId: string, root: string = process.cwd()) {
  return withoutLeadingSlash(
    moduleId
      .replace(root, '')
      .replace(/\/*node_modules\//, ''),
  )
}
