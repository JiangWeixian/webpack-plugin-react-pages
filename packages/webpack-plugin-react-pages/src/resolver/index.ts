export { nextEnhancedResolver } from './next-enhanced'
