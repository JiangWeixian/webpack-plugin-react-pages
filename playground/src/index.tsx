import { createRoot } from 'react-dom/client'
import { Provider } from 'react-redux'

import App from './App'
import { store } from './store'

const $ROOT = document.querySelector('#app')
const root = createRoot($ROOT)

const renderApp = (Component: any) => {
  root.render(
    <Provider store={store}>
      <Component />
    </Provider>,
  )
}

document.addEventListener('DOMContentLoaded', () => {
  renderApp(App)
})
