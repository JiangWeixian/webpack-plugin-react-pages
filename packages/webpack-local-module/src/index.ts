import { join } from 'node:path'

import { outputFileSync } from 'fs-extra'

import { resolveId } from './utils'

import type { Compiler, RspackPluginInstance } from '@rspack/core'

type Modules = Record<string, string>

interface ResolvedOptions {
  root: string
  moduleIdToPath: Record<string, string>
  moduleIdToContent: Record<string, string>
}

export interface Options {
  /**
   * @description Write virtual module into dirname
   * @default .rlm
   */
  dirname?: string
}

const DIRNAME = '.rlm'

export class WebpackLocalModule implements RspackPluginInstance {
  private modules: Modules = {}
  private options: Options = {}
  private resolvedOptions: ResolvedOptions = {} as ResolvedOptions
  constructor(modules: Modules, options: Options = {}) {
    this.modules = modules
    this.options = options
  }

  apply(compiler: Compiler) {
    this.resolvedOptions = this.resolveOptions(
      compiler,
      this.options,
      this.modules,
    )
    compiler.options.resolve.alias = {
      ...compiler.options.resolve.alias,
      ...(this.resolvedOptions.moduleIdToPath ?? {}),
    }
    this.writeModules(this.resolvedOptions.moduleIdToContent)
  }

  writeModule(moduleId: string, nextContent: string) {
    const id = resolveId(moduleId, this.resolvedOptions.root)
    const path = this.resolvedOptions.moduleIdToPath[id]
    if (path) {
      outputFileSync(path, nextContent)
    }
  }

  private writeModules(modules: Modules) {
    for (const [moduleId, content] of Object.entries(modules)) {
      this.writeModule(moduleId, content)
    }
  }

  private resolveOptions(compiler: Compiler, options: Options, modules: Record<string, string>) {
    const resolvedRoot = join(compiler.context ?? process.cwd(), 'node_modules', options.dirname ?? DIRNAME)
    const moduleIdToPath: Record<string, string> = {}
    const moduleIdToContent: Record<string, string> = {}
    for (const [moduleId] of Object.entries(modules)) {
      const id = resolveId(moduleId, resolvedRoot)
      moduleIdToPath[id] = join(resolvedRoot, id)
      moduleIdToContent[id] = this.modules[moduleId]
    }
    return {
      root: resolvedRoot,
      moduleIdToPath,
      moduleIdToContent,
    }
  }
}
