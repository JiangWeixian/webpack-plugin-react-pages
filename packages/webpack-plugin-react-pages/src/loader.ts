import { resolve } from 'node:path'

import { logger } from './utils'

import type webpack from 'webpack'
import type { Compiler } from './types'

interface RoutesLoaderOptions {
  namespace?: string
}

async function RoutesLoader(
  this: webpack.LoaderContext<RoutesLoaderOptions>,
  _source: string,
  ...args: any
) {
  const callback = this.async()
  const { namespace } = this.getOptions()
  this.cacheable(true)
  let $page = (this._compiler as Compiler).$page
  if (namespace) {
    $page = this._compiler?.[namespace]?.$page
  }
  // Make HMR work
  $page.options.dirs.forEach((options) => {
    const dir = resolve(this._compiler.context, options.dir)
    this.addContextDependency(dir)
  })

  logger('page instance', $page)

  $page.pageRouteMap.clear()
  await $page.searchGlob()
  const routes = await $page.resolveRoutes()
  logger('routes content', routes)

  callback(null, routes, ...args)
  return null
}

export default RoutesLoader
