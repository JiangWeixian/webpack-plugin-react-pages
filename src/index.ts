import type webpack from 'webpack'
import { NormalModuleReplacementPlugin } from 'webpack'
// eslint-disable-next-line import/no-extraneous-dependencies -- rollup will bundle this package
import { PageContext } from 'vite-plugin-pages'
import VirtualModulesPlugin from 'webpack-virtual-modules'
import { resolve } from 'pathe'
import { createFilteredWatchFileSystem } from './wfs'
import { logger } from './utils'

import { VIRTUAL_ROUTES_ID_TEST } from './constants'

const routesLoader = resolve(__dirname, 'loader.cjs')
const PLUGIN = 'WEBAPCK_PLUGIN_REACT_PAGES'

export class WebpackPluginReactPages {
  vm: VirtualModulesPlugin
  nmp: NormalModuleReplacementPlugin
  private _watchRunPatched: WeakSet<webpack.Compiler> = new WeakSet()
  constructor() {
    this.vm = new VirtualModulesPlugin({
      'virtual-react-pages.ts': `
      const routes = []
      export default routes;
      `,
    })
    this.nmp = new NormalModuleReplacementPlugin(/^virtual:react-pages/, (resource) => {
      resource.request = 'virtual-react-pages'
    })
  }

  apply(compiler: webpack.Compiler) {
    const page = new PageContext({ resolver: 'react', extensions: ['ts', 'tsx', 'js', 'jsx'] })
    if (!compiler.options.resolve) {
      compiler.options.resolve = {}
    }
    compiler.options.module.rules.push({
      include(resource) {
        return VIRTUAL_ROUTES_ID_TEST.test(resource)
      },
      enforce: 'pre',
      use: [
        {
          loader: routesLoader,
        },
      ],
    })

    // rollup typo
    compiler.hooks.compilation.tap(PLUGIN, async (compilation: any) => {
      for (const dir of page.options.dirs) {
        logger(dir)
        compilation.contextDependencies.add(resolve(compiler.context, dir.dir))
      }
    })

    // setup alias
    compiler.options.resolve.alias = {
      ...compiler.options.resolve.alias,
      'virtual-react-pages': resolve(compiler.context, 'virtual-react-pages.ts'),
    }

    // webpack-virtual-modules include wrong webpack types directly
    // Applying a webpack compiler to the virtual module
    this.vm.apply(compiler as any)
    this.nmp.apply(compiler)

    compiler.hooks.beforeCompile.tap(PLUGIN, async () => {
      // fs watcher unlink run after searchGlob
      // always generate new page route Map
      page.pageRouteMap.clear()
      await page.searchGlob()
      const routes = await page.resolveRoutes()
      logger('before compile')
      this.vm.writeModule('virtual-react-pages.ts', routes)
    })

    // related to pr: https://github.com/sysgears/webpack-virtual-modules/pull/129/files
    if (!this._watchRunPatched.has(compiler)) {
      compiler.watchFileSystem = createFilteredWatchFileSystem(compiler.watchFileSystem as any)
      this._watchRunPatched.add(compiler)
    }
  }
}
