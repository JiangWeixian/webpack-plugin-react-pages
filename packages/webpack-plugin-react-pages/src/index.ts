import { fileURLToPath } from 'node:url'

import {
  dirname,
  join,
  resolve,
} from 'pathe'
// eslint-disable-next-line import/no-extraneous-dependencies -- rollup will bundle this package
import { PageContext } from 'vite-plugin-pages'
import webpack from 'webpack'
import { WebpackLocalModule } from 'webpack-local-module'
import VirtualModulesPlugin from 'webpack-virtual-modules'

import { createFilteredWatchFileSystem } from './wfs'

import type { Options as WebpackLocalModuleOptions } from 'webpack-local-module'
import type { Compiler } from './types'
import type {
  PageContext as PageContextImpl,
  PageResolver,
  UserOptions,
} from './vite-plugin-pages-types'

const __dirname = dirname(fileURLToPath(import.meta.url))
const routesLoader = resolve(__dirname, 'loader.cjs')
const PLUGIN = 'WEBPACK_PLUGIN_REACT_PAGES'
const template = `
const routes = []
export default routes;
`

type WebpackPluginReactPagesOptions = Omit<
  UserOptions,
  // omit deprecated options, omit non-react-resolver only support react framework
  'moduleId' | 'nuxtStyle' | 'pagesDir' | 'replaceSquareBrackets' | 'resolver' | 'syncIndex'
> & {
  resolver?: PageResolver
  /**
   * @description Register PageContext instance namespace
   * if namespace defined, access via compiler[namespace].page, otherwise compiler.page
   * @default undefined
   */
  namespace?: string
  /**
   * @description With rspack bundler, virtual react pages module will write into disk
   * @default false
   */
  rspack?: boolean
  /**
   * @description With rspack bundler, use `webpack-local-module` custom behavior with localModuleOptions
   */
  localModuleOptions?: WebpackLocalModuleOptions
}

const isVirtualSchemaModule = (id: string) => id.includes('virtual:')

export class WebpackPluginReactPages {
  vm: VirtualModulesPlugin | WebpackLocalModule
  nmp?: webpack.NormalModuleReplacementPlugin
  page: PageContextImpl
  /**
   * @description Modules used in project
   */
  moduleIds: string[]
  /**
   * @description Resolved modules write via virtual-module
   * will replace `virtual:` to `virtual-`
   */
  resolvedModuleIds: string[]
  /**
   * @description Filter modules id include `virtual:`
   */
  virtualSchemaModuleIds: string[]
  moduleRE: RegExp
  resolvedModuleRE: RegExp
  shouldSupportVirtualModules = false
  namespace?: string
  rspack?: boolean
  private _watchRunPatched: WeakSet<Compiler> = new WeakSet()
  constructor({
    extensions = ['ts', 'tsx', 'js', 'jsx'],
    routeStyle = 'remix',
    namespace,
    rspack = false,
    localModuleOptions,
    ...options
  }: WebpackPluginReactPagesOptions = {}) {
    this.page = new PageContext({
      extensions,
      routeStyle,
      resolver: 'react',
      ...options,
      // TODO: type safe
    } as any) as any
    this.namespace = namespace
    this.moduleIds = this.page.options.resolver.resolveModuleIds()
    this.resolvedModuleIds = this.moduleIds.map((id) => {
      let resolvedId = id
      // virtual:react-pages -> virtual-react-pages
      if (isVirtualSchemaModule(id)) {
        this.shouldSupportVirtualModules = true
        resolvedId = id.replace('virtual:', 'virtual-')
      }
      return resolvedId
    })
    this.virtualSchemaModuleIds = this.moduleIds.filter(isVirtualSchemaModule)
    this.moduleRE = new RegExp(`(${this.moduleIds.join('|')})`)
    this.resolvedModuleRE = new RegExp(`(${this.resolvedModuleIds.join('|')})`)
    const modules = {}
    this.resolvedModuleIds.forEach((id) => {
      modules[`node_modules/${id}`] = template
    })
    this.rspack = rspack
    this.vm = rspack
      ? new WebpackLocalModule({
        ...modules,
      }, localModuleOptions ?? { dirname: '.react-routes' })
      : new VirtualModulesPlugin({
        ...modules,
      })
  }

  apply(compiler: Compiler) {
    const invalid = () => {
      this.resolvedModuleIds.forEach((id) => {
        this.vm.writeModule(`node_modules/${id}`, template)
      })
    }
    if (this.namespace) {
      compiler[this.namespace] = {}
      compiler[this.namespace].$page = this.page
      compiler[this.namespace].$page.invalid = invalid
    } else {
      compiler.$page = this.page
      compiler.$page.invalid = invalid
    }
    // support `virtual:` protocol in webpack@5
    if (this.shouldSupportVirtualModules && !this.rspack) {
      this.nmp = new webpack.NormalModuleReplacementPlugin(this.moduleRE, (resource) => {
        resource.request = resolve(
          compiler.context,
          `node_modules/${resource.request.replace('virtual:', 'virtual-')}`,
        )
      })
      this.nmp.apply(compiler)
    }

    if (!compiler.options.resolve) {
      compiler.options.resolve = {}
    }
    if (this.rspack) {
      compiler.options.resolve.alias = {
        ...compiler.options.resolve.alias,
        // rspack not support import file from `/src/*`
        '/src': join(this.page.options.root, 'src'),
      }
    }
    const resolvedModuleRE = this.resolvedModuleRE
    compiler.options.module.rules.push({
      include(resource) {
        return resolvedModuleRE.test(resource)
      },
      enforce: 'pre',
      use: [
        {
          loader: routesLoader,
          options: {
            namespace: this.namespace,
          },
        },
      ],
    })

    // webpack-virtual-modules include webpack@v4 types directly
    // Applying a webpack compiler to the virtual module
    this.vm.apply(compiler as any)

    // Register virtual module during build
    !this.rspack && compiler.hooks.compilation.tap(PLUGIN, () => {
      invalid()
    })

    // rspack not support webpack-virtual-modules, disable patch watch file system
    // related to pr: https://github.com/sysgears/webpack-virtual-modules/pull/129/files
    if (!this._watchRunPatched.has(compiler) && !this.rspack) {
      compiler.watchFileSystem = createFilteredWatchFileSystem(compiler.watchFileSystem as any)
      this._watchRunPatched.add(compiler)
    }
  }
}
