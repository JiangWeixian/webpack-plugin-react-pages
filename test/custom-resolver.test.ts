import { PageContext } from 'vite-plugin-pages'
import path from 'path'
import { describe, it } from 'vitest'

import { reactResolver } from '../src/resolver'

const fixtures = (...args: string[]) => path.resolve(__dirname, './fixtures', ...args)

describe('custom-resolver', () => {
  it('TODO', async () => {
    const page = new PageContext(
      {
        importMode: 'sync',
        routeStyle: 'remix',
        resolver: reactResolver() as any,
        dirs: ['react'],
      },
      fixtures(),
    )
    await page.searchGlob()
    const routes = await page.resolveRoutes()
    console.log(routes)
    expect(routes).toMatchSnapshot()
  })
})
