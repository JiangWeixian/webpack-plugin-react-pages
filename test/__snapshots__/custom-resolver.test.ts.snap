// Vitest Snapshot v1

exports[`custom-resolver > next enhanced resolver 1`] = `
"import React from \\"react\\";
import __pages_import_0__ from \\"/react/[...all].tsx\\";
import __pages_import_1__ from \\"/react/__auth.tsx\\";
import __pages_import_2__ from \\"/react/__auth/login.tsx\\";
import __pages_import_3__ from \\"/react/index.tsx\\";
import __pages_import_4__ from \\"/react/user.tsx\\";
import __pages_import_5__ from \\"/react/user/[id].tsx\\";
import __pages_import_6__ from \\"/react/user/index.tsx\\";

const routes = [{\\"caseSensitive\\":false,\\"path\\":\\"*\\",\\"element\\":React.createElement(__pages_import_0__)},{\\"caseSensitive\\":false,\\"element\\":React.createElement(__pages_import_1__),\\"children\\":[{\\"caseSensitive\\":false,\\"path\\":\\"login\\",\\"element\\":React.createElement(__pages_import_2__)}]},{\\"caseSensitive\\":false,\\"path\\":\\"/\\",\\"element\\":React.createElement(__pages_import_3__)},{\\"caseSensitive\\":false,\\"path\\":\\"user\\",\\"element\\":React.createElement(__pages_import_4__),\\"children\\":[{\\"caseSensitive\\":false,\\"path\\":\\":id\\",\\"element\\":React.createElement(__pages_import_5__)},{\\"caseSensitive\\":false,\\"path\\":\\"\\",\\"element\\":React.createElement(__pages_import_6__)}]}];

export default routes;"
`;
